packages/flutter_tools/test/general.shard/web/devfs_web_test.dart

```diff
@@ -103,12 +104,29 @@ void main() {
       .handleRequest(Request('GET', Uri.parse('http://foobar/foo.js')));
 
     expect(response.headers, allOf(<Matcher>[
-      containsPair('content-length', '9'),
-      containsPair('content-type', 'application/javascript'),
+      containsPair(HttpHeaders.contentLengthHeader, '9'),
+      containsPair(HttpHeaders.contentTypeHeader, 'application/javascript'),
+      containsPair(HttpHeaders.etagHeader, isNotNull),
+      containsPair(HttpHeaders.cacheControlHeader, 'max-age=0, must-revalidate')
     ]));
     expect((await response.read().toList()).first, utf8.encode('main() {}'));
   }));
 
+  test('Returns notModified when the ifNoneMatch header matches the etag', () => testbed.run(() async {
+    webAssetServer.writeFile('/foo.js', 'main() {}');
+
+    final Response response = await webAssetServer
+      .handleRequest(Request('GET', Uri.parse('http://foobar/foo.js')));
+    final String etag = response.headers[HttpHeaders.etagHeader];
+
+    final Response cachedResponse = await webAssetServer
+      .handleRequest(Request('GET', Uri.parse('http://foobar/foo.js'), headers: <String, String>{
+        HttpHeaders.ifNoneMatchHeader: etag
+      }));
+
+    expect(cachedResponse.statusCode, HttpStatus.notModified);
```

Comment:
nit: you might want to expect that the body is also empty

Reply:
I took a shot at this from the Github edit file interface, we'll see if I can write Dart without an IDE....